Add L-system trees as decorations

cx384 opened the pull request:
## Goal of the PR
Resolves #7346

## How does it work
It uses the same L-system as `minetest.spawn_tree`, but it's faster since it uses the engine version of `make_ltree`, which directly writes it into the VoxelManipulator during map generation and does not trigger a light update. 

## Roadmap
It's listed at https://dev.minetest.net/TODO.

## To do

This PR is Ready for Review.

## How to test

Add some Decorations, e.g. in MTG.
(`aaa` is a good seed for v7)
```lua
minetest.register_decoration({
	name = "default:testdeco",
	deco_type = "lsystem",
	place_on = {"default:dirt_with_grass", "default:sand"},
	sidelen = 80,
	fill_ratio = 0.005,
	y_max = 31000,
	y_min = 1,
	treedef = {
		axiom="FFFFFA",
		rules_a="[&&&F][&&&++++F][&&&----F]",
		trunk="default:coral_skeleton",
		leaves="default:permafrost_with_stones",
		angle=30,
		iterations=2,
		random_level=0,
		trunk_type="single",
		thin_branches=true,
	},
})
```

Also test if alias work:

- replace `default:coral_skeleton` with `alias_test1`,
- replace `default:permafrost_with_stones` with `alias_test2`
- and add:

```lua
minetest.register_alias("alias_test1", "default:sand")
minetest.register_alias("alias_test2", "default:snow")
```

Wuzzy2 commented:
In the documentation, I would remove the `treedef` example as its redundant. Referring to the `minetest.spawn_tree` function is enough.

cx384 commented:
I have removed the redundant example, but I'm unsure what should be there, since for all other fields there are proper examples in the table. However, a reference to the `L-system trees` section is probably enough.
Moreover, I added a features flag and rebased.

Wuzzy2 commented:
A question: Since L-system trees might have some randomness involved, how does this randomness get applied in the mapgen?

Does this guarantee that equal world seeds will generate an identical map with identical L-system trees? If not, that should be fixed. Because I think all mapgen stuff (terrain, biomes, decorations, etc.) should be equal if the world seed is the same.

cx384 commented:
> A question: Since L-system trees might have some randomness involved, how does this randomness get applied in the mapgen?

The position is calculated like for any other decoration, depending on the world seed, and the L-system seed gets calculated using the position. (Or the seed will be static if defined in the tree definition.)
https://github.com/minetest/minetest/blob/3cac17d23e2e0bfe74f1f627a4e206fec981fa4a/src/mapgen/treegen.cpp#L154

> Does this guarantee that equal world seeds will generate an identical map with identical L-system trees?
Yes.

The only concern may be, that two L-system decoration, which get placed at the same position in two different worlds, will look exactly the same. However, this does not happen too often and I doubt anyone would notice.

To fix this, the L-system tree seed could generally be calculated differently by taking the world seed into account, but this is out of scope for this PR.

cx384 commented:
Rebased.

SmallJoker commented on doc/lua_api.md:
```diff
@@ -4473,6 +4473,12 @@ Can specify a probability of a node randomly appearing when placed.
 This decoration type is intended to be used for multi-node sized discrete
 structures, such as trees, cave spikes, rocks, and so on.
 
+`lsystem`
+-----------
+Generates a L-system tree at the position where the decoration is placed.
+Uses the same L-system as `minetest.spawn_tree`, but is faster than using it manually.
+The `treedef` field in the decoration definition is used for the tree definition.
+
```
```suggestion
`lsystem`
-----------

Generates a L-system tree at the position where the decoration is placed.
Uses the same L-system as `minetest.spawn_tree`, but is faster than using it manually.
The `treedef` field in the decoration definition is used for the tree definition.
```

Same formatting as above.

SmallJoker commented on src/mapgen/mg_decoration.h:
```diff
@@ -112,12 +113,15 @@ class DecoSchematic : public Decoration {
+
+	virtual size_t generate(MMVManip *vm, PcgRandom *pr, v3s16 p, bool ceiling);
+
+	const NodeDefManager *ndef;
```
Please initialize to nullptr.

SmallJoker commented on src/mapgen/mg_decoration.h:
```diff
@@ -24,6 +24,7 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "objdef.h"
 #include "noise.h"
 #include "nodedef.h"
+#include "mapgen/treegen.h"
```
You could avoid including this header by using pointers (or references for that matter):

```
namespace treegen {
	struct TreeDef;
};

// .... in DecoLSystem:

treegen::TreeDef *tree_def = nullptr;
```

cx384 replied:
It now uses a shared pointer and `mapgen/treegen.h` does not have to be included anymore.
It's a shared pointer, since the `DecoLSystem` has a clone function. (I haven't looked up how often it will actually be cloned.)

I don't know if this it is better than storing the `TreeDef` inside the struct, but it is forward declared now, as you suggested.

SmallJoker reviewed (commented):
PR works what it promises. I noticed that some decorations might generate in the same place, which is ugly and needs fixing by someone in a separate PR later on.

The comments aside - looks good so far. Thank you.

sfan5 commented on src/script/lua_api/l_env.cpp:
```diff
@@ -1342,7 +1342,7 @@ int ModApiEnv::l_find_path(lua_State *L)
 }
 
-static bool read_tree_def(lua_State *L, int idx,
+bool ModApiEnvVM::read_tree_def(lua_State *L, int idx,
```
should be moved to `common/c_converter.{cpp,h}`

cx384 replied:
I moved to `c_content` and not `c_converter` since it's more fitting, I think.

sfan5 commented on src/script/lua_api/l_env.h:
```diff
@@ -22,6 +22,7 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "lua_api/l_base.h"
 #include "serverenvironment.h"
 #include "raycast.h"
+#include "mapgen/treegen.h"
```
forward declaration should work here without any extra changes

cx384 replied:
Completely removed the include.

sfan5 commented on src/mapgen/mg_decoration.h:
```diff
@@ -112,12 +113,15 @@ class DecoSchematic : public Decoration {
+
+	virtual size_t generate(MMVManip *vm, PcgRandom *pr, v3s16 p, bool ceiling);
+
+	const NodeDefManager *ndef;
```
also noticed something else, `Decoration` inherits `NodeResolver` so you already have `m_ndef`

cx384 replied:
Now it reuses the `NodeDefManager` from `NodeResolver`.

cx384 commented:
Thanks for the reviews, SmallJoker and sfan5.
I handled every suggestion, and rebased. (since it didn't compile with the Irrlicht version)
Note that
- `read_deco_lsystem` still needs a `ndef` as parameter since it can't access the one from `DecoLSystem`.
- `read_tree_def` moved to `c_content.{cpp,h}` and not `c_converter.{cpp,h}`, 
- and `tree_def` is a `shared_ptr` now.

SmallJoker commented on src/script/common/c_content.cpp:
```diff
@@ -2009,6 +2010,45 @@ void push_noiseparams(lua_State *L, NoiseParams *np)
+	getstringfield(L, idx, "rules_c", tree_def.rules_c);
+	getstringfield(L, idx, "rules_d", tree_def.rules_d);
+	getstringfield(L, idx, "trunk", trunk);
+	tree_def.trunknode = ndef->getId(trunk);
```
I know this code was not modified but the use-case is slightly different to `minetest.spawn_tree`. Decorations use the NodeResolver to properly deal with `unregister_item` and new aliases.
`TreeDef` would need to be like `Schematic` or `Decoration`: inherit `NodeResolver` and use `m_nodenames` + `m_nnlistsizes`.

cx384 replied:
Thanks for pointing this out, I completely missed it.
Now it uses a `NodeResolverr`, I hope I implemented it the right way.

(Also, I fixed the `place_on` behavior, maybe `canPlaceDecoration` should be called before any decoration function gets called.)

SmallJoker reviewed (approved):
Works on my machine.

sfan5 commented on src/mapgen/treegen.h:
```diff
@@ -35,7 +35,14 @@ namespace treegen {
+		// Needed since m_ndef can't be accessed in read_tree_def.
+		void setNodeDefManager(const NodeDefManager* ndef) {
+			NodeResolver::m_ndef = ndef;
+		}
```
you're using the noderesolver class wrong. admittedly there is no docs and I had to find out how it works myself.

1. do not set `m_ndef`
2. call `pendNodeResolve` after populating the lists
3. `resolveNodeNames` must be a virtual function (it is defined in `NodeResolver`)
4. do not manually call this function
5. you do not need an extra code path to handle the non-registration case. NodeResolver works transparently at runtime.

cx384 replied:
Thank you for the explanation, I didn't even know that `pendNodeResolve` existed. This makes total sense.

I fixed it. Now the node resolve things are only in `read_tree_def` and in `TreeDef.resolveNodeNames()`, such that only the `NodeDefManager` has to handle it.

sfan5 commented on src/script/lua_api/l_env.h:
```diff
@@ -245,7 +245,7 @@ class ModApiEnv : public ModApiEnvBase {
 /*
  * Duplicates of certain env APIs that operate not on the global
- * map but on a VoxelManipulator. This is for emerge scripting. 
+ * map but on a VoxelManipulator. This is for emerge scripting.
```
stray whitespace change

cx384 replied:
Fixed

sfan5 commented on src/mapgen/mg_decoration.cpp:
```diff
@@ -472,3 +473,27 @@ size_t DecoSchematic::generate(MMVManip *vm, PcgRandom *pr, v3s16 p, bool ceilin
+{
+	if (!canPlaceDecoration(vm, p))
+		return 0;
+	return treegen::make_ltree(*vm, p, m_ndef, *tree_def);
```
to make sure it's definitely not used as non-const I'd do this:
```suggestion
	// shared
	const auto &ref = *tree_def;
	return treegen::make_ltree(*vm, p, m_ndef, ref);
```

sfan5 commented on src/mapgen/treegen.cpp:
```diff
@@ -32,6 +32,16 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 
+void TreeDef::resolveNodeNames()
+{
+	getIdFromNrBacklog(&trunknode.param0, "", CONTENT_IGNORE, false);
+	getIdFromNrBacklog(&leavesnode.param0, "", CONTENT_IGNORE, false);
+	if(leaves2_chance)
+		getIdFromNrBacklog(&leaves2node.param0, "", CONTENT_IGNORE, false);
+	if(fruit_chance)
+		getIdFromNrBacklog(&fruitnode.param0, "", CONTENT_IGNORE, false);
```
error_on_fallback=true
```suggestion
	getIdFromNrBacklog(&trunknode.param0, "", CONTENT_IGNORE);
	getIdFromNrBacklog(&leavesnode.param0, "", CONTENT_IGNORE);
	if (leaves2_chance)
		getIdFromNrBacklog(&leaves2node.param0, "", CONTENT_IGNORE);
	if (fruit_chance)
		getIdFromNrBacklog(&fruitnode.param0, "", CONTENT_IGNORE);
```

sfan5 commented on src/mapgen/mg_decoration.cpp:
```diff
@@ -472,3 +473,24 @@ size_t DecoSchematic::generate(MMVManip *vm, PcgRandom *pr, v3s16 p, bool ceilin
+	if (!canPlaceDecoration(vm, p))
+		return 0;
+
+	// Make sure that tree_def can't be overridden, since it is shared.
```
```suggestion
	// Make sure that tree_def can't be modified, since it is shared.
```

sfan5 commented on src/mapgen/treegen.cpp:
```diff
@@ -32,6 +32,16 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 
+void TreeDef::resolveNodeNames()
+{
+	getIdFromNrBacklog(&trunknode.param0, "", CONTENT_UNKNOWN);
```
i don't see `CONTENT_UNKNOWN` being used in this role in the engine at all.
not even sure what happens when you manually place this.

sfan5 reviewed (approved):
works
